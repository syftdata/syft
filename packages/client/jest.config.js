/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  globalSetup: './tests/jest-setup.js'
};
