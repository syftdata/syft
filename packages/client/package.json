{
  "name": "@syftdata/client",
  "version": "0.1.33-alpha.6",
  "description": "Client library to log events. Please use syft to generate events from a schema file.",
  "keywords": [
    "syft",
    "client",
    "events",
    "data",
    "instrumentation",
    "analytics"
  ],
  "author": "Nageswara Rao Mannem <nageswara@syftdata.com>",
  "homepage": "https://syftdata.com",
  "license": "Apache-2.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "directories": {
    "src": "./src",
    "test": "./tests"
  },
  "files": [
    "./lib",
    "README.md",
    "LICENSE"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/syftdata/syft.git"
  },
  "bugs": {
    "url": "https://github.com/syftdata/syft/issues"
  },
  "browser": {
    "./src/node/configstore.js": "./src/browser/configstore.js",
    "./src/node/dispatcher.js": "./src/browser/dispatcher.js",
    "./src/node/uploader.js": "./src/browser/uploader.js",
    "./src/node/index.js": "./src/browser/index.js"
  },
  "scripts": {
    "prebuild": "rm -rf ./lib",
    "build": "tsc && tsc --project tsconfig_scripts.json",
    "build-bundle": "npx webpack --config webpack.config.js",
    "test": "jest",
    "lint": "eslint --ext .ts src tests --fix",
    "prepublish": "npm run build",
    "postinstall": "node ./lib/scripts/postinstall.js || true",
    "prettier-format": "prettier --config .prettierrc 'src/**/*.ts' --write"
  },
  "devDependencies": {
    "@types/jest": "^29.5.4",
    "@types/js-cookie": "^3.0.3",
    "@types/node": "^18.13.0",
    "@typescript-eslint/eslint-plugin": "^5.51.0",
    "@typescript-eslint/parser": "^5.51.0",
    "eslint": "^8.36.0",
    "eslint-config-prettier": "^8.6.0",
    "eslint-config-standard-with-typescript": "^34.0.0",
    "eslint-plugin-import": "^2.27.5",
    "eslint-plugin-n": "^15.6.1",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-promise": "^6.1.1",
    "jest": "^29.6.3",
    "jest-environment-jsdom": "^29.6.3",
    "prettier": "^2.8.4",
    "ts-jest": "^29.0.5",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "js-cookie": "^3.0.5",
    "uuid": "^9.0.0",
    "zod": "^3.20.5"
  },
  "gitHead": "88c566d72818936b193d860ced18df01b84bf140"
}
