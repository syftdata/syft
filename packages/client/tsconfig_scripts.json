{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./lib/scripts",
    "lib": ["ES2015"]
  },
  "include": [
    "./scripts"
  ]
}