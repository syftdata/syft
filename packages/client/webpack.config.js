const path = require('path');
// const WebpackBundleAnalyzer = require('webpack-bundle-analyzer');

const generalConfig = {
  mode: 'production',
  resolve: {
    extensions: ['.ts']
  },
  plugins: [
    //  new WebpackBundleAnalyzer.BundleAnalyzerPlugin()
  ],
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: [/node_modules/, /lib/, /lib-bundle/, /dist/, /lib-esm/]
      }
    ]
  }
};

const nodeConfig = {
  target: 'node',
  output: {
    filename: '[name].node.js',
    path: path.resolve(__dirname, 'lib-bundle'),
    library: {
      type: 'module'
    }
  }
};

const browserConfig = {
  target: 'web',
  output: {
    filename: '[name].web.js',
    path: path.resolve(__dirname, 'lib-bundle'),
    library: {
      type: 'module'
    }
  }
};

module.exports = (env, argv) => {
  if (argv.mode === 'development') {
    generalConfig.mode = 'development';
    generalConfig.devtool = 'source-map';
  }
  Object.assign(nodeConfig, generalConfig);
  Object.assign(browserConfig, generalConfig);
  return [nodeConfig, browserConfig];
};
