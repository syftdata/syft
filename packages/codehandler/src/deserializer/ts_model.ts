import { Project, ScriptTarget } from 'ts-morph';
import { logDetail, logError, logFatal } from '@syftdata/common/lib/utils';
import {
  type InputSource,
  type AST,
  type Sink
} from '@syftdata/common/lib/types';
import { getEventSchemas } from './visitor';
import * as fs from 'fs';
import * as path from 'path';
import {
  getConfigExpressionForProject,
  getConfigFromExpression,
  getInputs,
  getSinks
} from './config';

// TODO: memoize this method.
export function createTSProject(filePaths: string[]): Project {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5, strictNullChecks: true }
  });

  // if directory, add glob style
  const files = filePaths.map((filePath) => {
    if (fs.existsSync(filePath) && fs.lstatSync(filePath).isDirectory()) {
      return path.join(filePath, '**', '!(*.d).ts');
    }
    return filePath;
  });
  if (files.length > 0) {
    project.addSourceFilesAtPaths(files);
  }
  return project;
}

export function readTSProject(
  sourceDir: string
): { project: Project; packageJson: any } | undefined {
  const tsConfigFilePath = `${sourceDir}/tsconfig.json`;
  if (!fs.existsSync(tsConfigFilePath)) {
    logError(
      `tsconfig.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read package.json file.
  const packageJsonPath = `${sourceDir}/package.json`;
  if (!fs.existsSync(packageJsonPath)) {
    logError(
      `package.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read all files in the project from source dir. and initialize ts-morph project
  const project = new Project({
    tsConfigFilePath
  });
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  return {
    project,
    packageJson
  };
}

export function generateASTForProject(project: Project): AST | undefined {
  const sourceFile = project.getSourceFile('events.ts');
  if (sourceFile == null) {
    logFatal(
      ':warning: events.ts file is not found. Please initialize your project using `syft init`'
    );
    return;
  }
  const relPath = path.relative('.', sourceFile.getFilePath());
  logDetail(`:heavy_check_mark: Parsing ${relPath}..`);
  const eventSchemas = getEventSchemas(project, sourceFile);

  const syftConfig = getConfigExpressionForProject(project);
  if (syftConfig === undefined) {
    logFatal(':warning: Config file is not found. Please provide one.');
    return;
  }
  const config = getConfigFromExpression(syftConfig);
  const sinks: Sink[] = getSinks(project);
  const inputs: InputSource[] = getInputs(project);
  return {
    eventSchemas,
    syftConfig,
    config,
    sinks,
    inputs
  };
}

export function deserialize(fileNames: string[]): AST | undefined {
  return generateASTForProject(createTSProject(fileNames));
}

export function deserializeFromCode(
  code: string,
  config: string
): AST | undefined {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5 }
  });
  project.createSourceFile(path.join('events.ts'), (writer) => {
    writer.write(code);
  });
  project.createSourceFile(path.join('config.ts'), (writer) => {
    writer.write(config);
  });
  return generateASTForProject(project);
}
