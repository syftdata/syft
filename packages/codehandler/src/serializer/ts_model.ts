import { type AST } from '@syftdata/common/lib/types';
import * as handlebars from 'handlebars/dist/cjs/handlebars';
import { EVENT_MODELS_TEMPLATE } from '../templates/model_template';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { isEmptyString } from '@syftdata/common/lib/utils';

// TODO: Memoize
function getEventModelsTemplate(): handlebars.HandlebarsTemplateDelegate<any> {
  return handlebars.compile(EVENT_MODELS_TEMPLATE, { noEscape: true });
}

export function serialize(ast: AST): string {
  const schemas = ast.eventSchemas.map((schema) => ({
    ...schema,
    fields: schema.fields.map((field) => ({
      ...field,
      has_documentation: !isEmptyString(field.documentation)
    })),
    eventType: SyftEventType[schema.eventType]
  }));
  return getEventModelsTemplate()({
    schemas
  });
}
