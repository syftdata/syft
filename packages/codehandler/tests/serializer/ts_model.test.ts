import { SyftEventType } from '@syftdata/client';
import { serialize } from '../../src/serializer/ts_model';

describe('serialize', () => {
  it('serialize regular primitive and syft types', async () => {
    const ast = {
      eventSchemas: [
        {
          name: 'TestEvent',
          eventType: SyftEventType.TRACK,
          documentation: 'This is a test event with two fields',
          fields: [
            {
              name: 'test',
              type: {
                name: 'string',
                zodType: 'z.string()'
              },
              documentation: 'This is a test field',
              isOptional: false
            },
            {
              name: 'email',
              type: {
                name: 'string',
                syfttype: 'Email',
                zodType: 'z.string().email()'
              },
              isOptional: false
            }
          ],
          exported: true,
          zodType: ''
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      }
    };
    expect(serialize(ast)).toMatchSnapshot();
  });
});
