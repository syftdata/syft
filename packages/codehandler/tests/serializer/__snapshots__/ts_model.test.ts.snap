// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize serialize regular primitive and syft types 1`] = `
"// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type { type, SyftEventType } from '@syftdata/client';

/**
 * This is a test event with two fields
 * @type {SyftEventType.TRACK}
 */
export class TestEvent {
  /**
   * This is a test field
   */
  test: string;

  email: type.Email;

}
"
`;
